win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+        /*Codes_SRS_FILE_43_007: [ file_destroy shall close the file handle handle. ]*/
+        /*Codes_SRS_FILE_WIN32_43_016: [ file_destroy shall call CloseHandle on handle->h_file ]*/
+        if (!CloseHandle(handle->h_file))
+        {
+            LogError("failure in CloseHandle");
+        }
+        /*Codes_SRS_FILE_WIN32_43_015: [ file_destroy shall close the threadpool IO by calling CloseThreadPoolIo. ]*/
+        CloseThreadpoolIo(handle->ptp_io);
+        /*Codes_SRS_FILE_WIN32_43_042: [ file_destroy shall free the handle.]*/
+        free(handle);
+    }
+    else
+    {
+        LogError("Invalid argument to file_destroy: FILE_HANDLE=%p", handle);
+    }
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, const unsigned char*, source, uint32_t, size, uint64_t, position, FILE_CB, user_callback, void*, user_context)
+{
+    FILE_WRITE_ASYNC_RESULT result;
+    if
+    (
+        /*Codes_SRS_FILE_43_009: [ If handle is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_043: [ If handle is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS.]*/
+        (handle == NULL) ||
+        /*Codes_SRS_FILE_43_010: [ If source is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_044: [ If source is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS.]*/
+        (source == NULL) ||
+        /*Codes_SRS_FILE_43_012: [ If user_callback is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_045: [ If user_callback is NULL then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS.]*/
+        (user_callback == NULL)||
+        /*Codes_SRS_FILE_43_040: [ If position + size is greater than INT64_MAX, then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_060: [ If position + size is greater than INT64_MAX, then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        ((position + size) > INT64_MAX)||
+        /*Codes_SRS_FILE_43_042: [ If size is 0 then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_061: [ If size is 0 then file_write_async shall fail and return FILE_WRITE_ASYNC_INVALID_ARGS. ]*/
+        (size == 0)
+    )
+    {
+        LogError("Invalid arguments to file_write_async: FILE_HANDLE file_handle=%p, const unsigned char* source=%p, uin32_t size=%" PRIu32 ", uint64_t position=%" PRIu64 ", FILE_CB user_callback=%p, void* user_context=%p",
+            handle, source, size, position, user_callback, user_context);
+        result = FILE_WRITE_ASYNC_INVALID_ARGS;
+    }
+    else
+    {
+        bool callback_will_be_called = false;

```

Comment:
>callback_will_be_called [](start = 13, length = 23)

This is more like the typical style in the codebase :) - a result variable and maybe a helper flag like this for cleanup.